Add in an event to retrieve keyboard events as a string

Currently there is no way to easily get key pressed events as a String. Conrod has this feature for example.

Hello,

I've forked the repository and done some work on this issue, I want to open a PR but first I'd like to ask two questions:
1. Is this something that is still desirable, seeing the issue was opened more than an year ago?
2. At the moment there is an error when running the script which checks whether the examples have had any regressions, I've seen that this same error is present in, at least, one of the most recent three actions runs on the repo. So should I wait for this error to be resolved or should I open my PR?

PS: The error in my second question is the following:
`error[E0599]: no function or associated item named `new_immediate` found for struct `FsEventWatcher` in the current scope
   --> /Users/r3xus/.cargo/git/checkouts/hotglsl-4b7ab76e967c6433/0066dea/src/lib.rs:174:51
    |
174 |     let mut watcher = notify::RecommendedWatcher::new_immediate(move |res| {
    |                                                   ^^^^^^^^^^^^^ function or associated item not found in `FsEventWatcher`
`

--- a/examples/nannou_basics/all_functions.rs
+++ b/examples/nannou_basics/all_functions.rs
@@ -62,6 +62,7 @@ fn window_event(_app: &App, _model: &mut Model, event: WindowEvent) {
     match event {
         KeyPressed(_key) => {}
         KeyReleased(_key) => {}
+        ReceivedCharacter(_char) => {}
         MouseMoved(_pos) => {}
         MousePressed(_button) => {}
         MouseReleased(_button) => {}

--- a/examples/offline/tree.rs
+++ b/examples/offline/tree.rs
@@ -105,6 +105,7 @@ fn window_event(_app: &App, _model: &mut Model, event: WindowEvent) {
     match event {
         KeyPressed(_key) => {}
         KeyReleased(_key) => {}
+        ReceivedCharacter(_char) => {}
         MouseMoved(_pos) => {}
         MousePressed(_button) => {}
         MouseReleased(_button) => {}

--- a/examples/rust_basics/1_nannou_events.rs
+++ b/examples/rust_basics/1_nannou_events.rs
@@ -31,6 +31,7 @@ fn event(_app: &App, _model: &mut Model, event: WindowEvent) {
         // Keyboard events
         KeyPressed(_key) => {}
         KeyReleased(_key) => {}
+        ReceivedCharacter(_char) => {}
 
         // Mouse events
         MouseMoved(_pos) => {}

--- a/guide/src/changelog.md
+++ b/guide/src/changelog.md
@@ -19,6 +19,7 @@ back to the origins.
   demonstration.
 - Fix audio examples on Windows by adding `WindowBuilder` method `windowsos_drag_and_drop`. Disabling this allows `cpal` and `winit` to coexist on the same thread.
 - Add community tutorials section to the guide.
+- Provided an user-friendly way to get the character value of a pressed keyboard key.
 
 ---
 

--- a/nannou/src/app.rs
+++ b/nannou/src/app.rs
@@ -1642,6 +1642,9 @@ where
             match simple {
                 event::WindowEvent::KeyPressed(key) => call_user_function!(key_pressed, key),
                 event::WindowEvent::KeyReleased(key) => call_user_function!(key_released, key),
+                event::WindowEvent::ReceivedCharacter(char) => {
+                    call_user_function!(received_character, char)
+                }
                 event::WindowEvent::MouseMoved(pos) => call_user_function!(mouse_moved, pos),
                 event::WindowEvent::MousePressed(button) => {
                     call_user_function!(mouse_pressed, button)

--- a/nannou/src/event.rs
+++ b/nannou/src/event.rs
@@ -121,6 +121,9 @@ pub enum WindowEvent {
     /// The given keyboard key was released.
     KeyReleased(Key),
 
+    /// Character input received event.
+    ReceivedCharacter(char),
+
     /// The mouse moved to the given x, y position.
     MouseMoved(Point2),
 
@@ -291,12 +294,15 @@ impl WindowEvent {
                 None => return None,
             },
 
+            winit::event::WindowEvent::ReceivedCharacter(char) => {
+                WindowEvent::ReceivedCharacter(char.clone())
+            }
+
             winit::event::WindowEvent::ModifiersChanged(_) => {
                 return None;
             }
 
             winit::event::WindowEvent::AxisMotion { .. }
-            | winit::event::WindowEvent::ReceivedCharacter(_)
             | winit::event::WindowEvent::ThemeChanged(_)
             | winit::event::WindowEvent::ScaleFactorChanged { .. } => {
                 return None;

--- a/nannou/src/window.rs
+++ b/nannou/src/window.rs
@@ -59,6 +59,7 @@ pub(crate) struct UserFunctions {
     pub(crate) raw_event: Option<RawEventFnAny>,
     pub(crate) key_pressed: Option<KeyPressedFnAny>,
     pub(crate) key_released: Option<KeyReleasedFnAny>,
+    pub(crate) received_character: Option<ReceivedCharacterFnAny>,
     pub(crate) mouse_moved: Option<MouseMovedFnAny>,
     pub(crate) mouse_pressed: Option<MousePressedFnAny>,
     pub(crate) mouse_released: Option<MouseReleasedFnAny>,
@@ -110,6 +111,9 @@ pub type KeyPressedFn<Model> = fn(&App, &mut Model, Key);
 
 /// A function for processing key release events.
 pub type KeyReleasedFn<Model> = fn(&App, &mut Model, Key);
+
+/// A function for processing received characters.
+pub type ReceivedCharacterFn<Model> = fn(&App, &mut Model, char);
 
 /// A function for processing mouse moved events.
 pub type MouseMovedFn<Model> = fn(&App, &mut Model, Point2);
@@ -209,6 +213,7 @@ fn_any!(EventFn<M>, EventFnAny);
 fn_any!(RawEventFn<M>, RawEventFnAny);
 fn_any!(KeyPressedFn<M>, KeyPressedFnAny);
 fn_any!(KeyReleasedFn<M>, KeyReleasedFnAny);
+fn_any!(ReceivedCharacterFn<M>, ReceivedCharacterFnAny);
 fn_any!(MouseMovedFn<M>, MouseMovedFnAny);
 fn_any!(MousePressedFn<M>, MousePressedFnAny);
 fn_any!(MouseReleasedFn<M>, MouseReleasedFnAny);
@@ -508,6 +513,14 @@ impl<'app> Builder<'app> {
         M: 'static,
     {
         self.user_functions.key_released = Some(KeyReleasedFnAny::from_fn_ptr(f));
+        self
+    }
+
+    pub fn received_character<M>(mut self, f: ReceivedCharacterFn<M>) -> Self
+    where
+        M: 'static,
+    {
+        self.user_functions.received_character = Some(ReceivedCharacterFnAny::from_fn_ptr(f));
         self
     }
 

